File: DesignPattern/creational_pattern/prototype_pattern.cpp
#include <iostream>
#include <string>

class Prototype {
public:
    virtual Prototype* clone() const = 0;

    virtual void setInfo(const std::string& info) {
        info_ = info;
    }

    virtual std::string getInfo() const {
        return info_;
    }

protected:
    std::string info_;
};

class ConcretePrototype : public Prototype {
public:
    Prototype* clone() const override {
        ConcretePrototype* newPrototype = new ConcretePrototype();
        newPrototype->setInfo(this->getInfo());
        return newPrototype;
    }
};

int main() {
    ConcretePrototype original;
    original.setInfo("This is the original object.");

    ConcretePrototype* cloned1 = dynamic_cast<ConcretePrototype*>(original.clone());
    cloned1->setInfo("This is the first cloned object.");

    ConcretePrototype* cloned2 = dynamic_cast<ConcretePrototype*>(original.clone());
    cloned2->setInfo("This is the second cloned object.");

    std::cout << "Original: " << original.getInfo() << std::endl;
    std::cout << "Cloned 1: " << cloned1->getInfo() << std::endl;
    std::cout << "Cloned 2: " << cloned2->getInfo() << std::endl;

    delete cloned1;
    delete cloned2;

    return 0;
}



/*
- The Prototype Pattern is a creational design pattern that allows you to create new objects by copying an existing object, known as the prototype. 
This pattern is useful when you want to create new objects with the same initial state as an existing object, saving the cost of initializing the new object from scratch.

- In this example: Prototype is the base class for the prototype objects. It defines a clone method, which will be implemented by concrete prototypes. 
The setInfo and getInfo methods allow you to set and retrieve the object's information.

- ConcretePrototype is a concrete class that inherits from Prototype. It implements the clone method, making a deep copy of the object, and also allows 
setting and getting the object's information.

- In the main function, we create an original object and then create two clones from it. Each clone is created by calling the clone method on the 
original object. We then modify the information for each object to demonstrate that they are distinct.

- After using the cloned objects, remember to delete them to free the allocated memory.

- The Prototype Pattern is useful when you want to create new objects that are similar to existing ones. It allows you to create objects with the same 
initial state without the need to reinitialize them.
*/

File: DesignPattern/creational_pattern/prototype_pattern.py
import copy

class Prototype:
    def clone(self):
        return copy.deepcopy(self)

    def set_info(self, info):
        self.info = info

    def get_info(self):
        return self.info

class ConcretePrototype(Prototype):
    def __init__(self):
        self.info = "Default information"

if __name__ == "__main__":
    original = ConcretePrototype()
    original.set_info("This is the original object.")

    cloned1 = original.clone()
    cloned1.set_info("This is the first cloned object.")

    cloned2 = original.clone()
    cloned2.set_info("This is the second cloned object.")

    print("Original:", original.get_info())
    print("Cloned 1:", cloned1.get_info())
    print("Cloned 2:", cloned2.get_info())



'''
Output:

Original: This is the original object.
Cloned 1: This is the first cloned object.
Cloned 2: This is the second cloned object.
'''


'''
- The Prototype Pattern in Python allows you to create new objects by copying an existing object, known as the prototype. This pattern is useful when you 
want to create new objects with the same initial state as an existing object, saving the cost of initializing the new object from scratch.

- In this example: Prototype is the base class for the prototype objects. It defines a clone method, which creates a deep copy of the object. The set_info and 
get_info methods allow you to set and retrieve the object's information.
- ConcretePrototype is a concrete class that inherits from Prototype. It sets a default value for the info attribute.

- In the main section, we create an original object and then create two clones from it. Each clone is created by calling the clone method on the original object. 
We then modify the information for each object to demonstrate that they are distinct.

- The Prototype Pattern is useful when you want to create new objects that are similar to existing ones. It allows you to create objects with the same initial 
state without the need to reinitialize them. In Python, the copy module is used to create deep copies of objects, ensuring that  the cloned objects are 
distinct from the original.
'''